File: app-java/src/main/java/com/tritiumstudios/phasmophobiaevidencetool/android/data/utilities/BitmapUtils.java
package com.tritiumstudios.phasmophobiaevidencetool.android.data.utilities;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.drawable.Drawable;
import android.util.Log;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;

import javax.microedition.khronos.egl.EGL10;
import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.egl.EGLContext;
import javax.microedition.khronos.egl.EGLDisplay;

/**
 * BitmapUtils class
 *
 * @author TritiumGamingStudios
 */
public class BitmapUtils {

    private int currentLayer = -1;

    private int maxTextureSize;
    public ArrayList<Integer> resources = new ArrayList<>();
    @NonNull
    public ArrayList<PorterDuff.Mode> filters = new ArrayList<>();

    /**
     *
     */
    public BitmapUtils() {
        setMaxTextureSize();
    }

    /**
     * @param resource
     * @return
     */
    @NonNull
    public BitmapUtils setResource(@DrawableRes int resource) {
        clearResources();
        addResource(resource);

        return this;
    }

    /**
     * @param resources
     */
    public void setResources(ArrayList<Integer> resources) {
        this.resources = resources;
    }

    /**
     * @param resource
     */
    @NonNull
    public BitmapUtils addResource(@DrawableRes int resource) {
        this.resources.add(resource);
        this.filters.add(null);

        return this;
    }

    /**
     * @param resource
     */
    @NonNull
    public BitmapUtils addResource(@DrawableRes int resource, PorterDuff.Mode filterMode) {
        this.resources.add(resource);
        this.filters.add(filterMode);

        return this;
    }

    /**
     *
     */
    public void clearResources() {
        resources = new ArrayList<>();
        filters = new ArrayList<>();
        currentLayer = -1;
    }

    /**
     *
     */
    private void setMaxTextureSize() {
        // Safe minimum default size
        final int IMAGE_MAX_BITMAP_DIMENSION = 2048;

        // Get EGL Display
        EGL10 egl = (EGL10) EGLContext.getEGL();
        EGLDisplay display = egl.eglGetDisplay(EGL10.EGL_DEFAULT_DISPLAY);

        // Initialise
        int[] version = new int[2];
        egl.eglInitialize(display, version);

        // Query total number of configurations
        int[] totalConfigurations = new int[1];
        egl.eglGetConfigs(display, null, 0, totalConfigurations);

        // Query actual list configurations
        EGLConfig[] configurationsList = new EGLConfig[totalConfigurations[0]];
        egl.eglGetConfigs(
                display,
                configurationsList,
                totalConfigurations[0],
                totalConfigurations);

        int[] textureSize = new int[1];
        int maximumTextureSize = 0;

        // Iterate through all the configurations to located the maximum texture size
        for (int i = 0; i < totalConfigurations[0]; i++) {
            // Only need to check for width since opengl textures are always squared
            egl.eglGetConfigAttrib(
                    display,
                    configurationsList[i],
                    EGL10.EGL_MAX_PBUFFER_WIDTH,
                    textureSize);

            // Keep track of the maximum texture size
            if (maximumTextureSize < textureSize[0]) {
                maximumTextureSize = textureSize[0];
            }
        }
        // Release
        egl.eglTerminate(display);

        maxTextureSize = Math.max(maximumTextureSize, IMAGE_MAX_BITMAP_DIMENSION);
    }

    /**
     * @param context
     * @return
     */
    @Nullable
    public Bitmap compileBitmaps(@NonNull Context context) {

        if(resources == null || filters == null) {
             return null;
        }

        Bitmap bitmap = null;
        //@DrawableRes int resource : resources

        for (int i = 0; i < resources.size(); i++) {
            //bitmap = createBitmap(context, bitmap, resource);
            int drawableRes = resources.get(i);
            PorterDuff.Mode filter = filters.get(i);

            bitmap = createBitmap(context, bitmap, drawableRes, filter);
        }
        return bitmap;

    }

    /**
     * @param context
     * @param previousBitmap
     * @return
     */
    public Bitmap compileNextBitmap(@NonNull Context context, @NonNull Bitmap previousBitmap) {
        currentLayer++;
        return createBitmap(context, previousBitmap, resources.get(currentLayer), null);
    }

    /**
     * @return
     */
    public boolean hasNextBitmap() {
        return currentLayer < resources.size() - 1;
    }

    /**
     * @param c
     * @param baseLayer
     * @param id
     * @return
     */
    /*
    private Bitmap createBitmap(Context c, Bitmap baseLayer, int id) {

        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inSampleSize = 1;
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(c.getResources(), id, options);

        // Raw height and width of image with respect to starting sampleSize
        final int height = options.outHeight, width = options.outWidth;
        //Get biggest image dimension between both width and height
        int highestDim = Math.max(height, width);
        double dimScale = (double) maxTextureSize / (double) highestDim * options.inDensity;
        if (dimScale < 1) {
            options.inSampleSize += (int) Math.ceil(Math.abs(dimScale));
        }
        options.inJustDecodeBounds = false;

        Bitmap bitmapToAdd = BitmapFactory.decodeResource(c.getResources(), id, options);
        if(bitmapToAdd == null) {
            bitmapToAdd = getBitmapFromVector(c, id, options);
        }

        return addLayer(baseLayer, bitmapToAdd);
    }
    */

    /**
     * @param c
     * @param baseLayer
     * @param id
     * @return
     */
    private Bitmap createBitmap(@NonNull Context c, @NonNull Bitmap baseLayer, int id, PorterDuff.Mode mode) {

        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inSampleSize = 1;
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(c.getResources(), id, options);

        // Raw height and width of image with respect to starting sampleSize
        final int height = options.outHeight, width = options.outWidth;
        //Get biggest image dimension between both width and height
        int highestDim = Math.max(height, width);
        double dimScale = (double) maxTextureSize / (double) highestDim * options.inDensity;
        if (dimScale < 1) {
            options.inSampleSize += (int) Math.ceil(Math.abs(dimScale));
        }
        options.inJustDecodeBounds = false;

        Bitmap bitmapToAdd = BitmapFactory.decodeResource(c.getResources(), id, options);
        if(bitmapToAdd == null) {
            //Temp Options
            BitmapFactory.Options optionsTemp = new BitmapFactory.Options();
            optionsTemp.inSampleSize = 1;
            optionsTemp.inJustDecodeBounds = true;
            optionsTemp.outWidth = baseLayer.getWidth();
            optionsTemp.outHeight = baseLayer.getHeight();
            optionsTemp.inSampleSize = options.inSampleSize;
            optionsTemp.inJustDecodeBounds = false;

            bitmapToAdd = getBitmapFromVector(c, id, optionsTemp);
        }

        return addLayer(baseLayer, bitmapToAdd, mode);
    }

    /**
     * @param baseLayer
     * @param topLayer
     * @return
     * @throws OutOfMemoryError
     */
    /*
    private Bitmap addLayer(Bitmap baseLayer, Bitmap topLayer) throws OutOfMemoryError {
        if (baseLayer == null && BitmapUtils.bitmapExists(topLayer)) {
            baseLayer = Bitmap.createBitmap(
                    topLayer.getWidth(),
                    topLayer.getHeight(),
                    topLayer.getConfig());
        }
        if (baseLayer != null && !baseLayer.isRecycled()) {
            Canvas canvas = new Canvas(baseLayer);
            if (BitmapUtils.bitmapExists(topLayer)) {
                canvas.drawBitmap(topLayer, new Matrix(), null);
                topLayer.recycle();
            }
        }
        System.gc();

        return baseLayer;
    }
    */

    /**
     * @param baseLayer
     * @param topLayer
     * @return
     * @throws OutOfMemoryError
     */
    @Nullable
    private Bitmap addLayer(@Nullable Bitmap baseLayer, @NonNull Bitmap topLayer, @Nullable PorterDuff.Mode mode) throws OutOfMemoryError {
        if (baseLayer == null && BitmapUtils.bitmapExists(topLayer)) {
            baseLayer = Bitmap.createBitmap(
                    topLayer.getWidth(),
                    topLayer.getHeight(),
                    topLayer.getConfig());
        }
        if (baseLayer != null && !baseLayer.isRecycled()) {
            Canvas canvas = new Canvas(baseLayer);
            if (BitmapUtils.bitmapExists(topLayer)) {
                Paint paint = null;
                if(mode != null) {
                    paint = new Paint();
                    paint.setXfermode(new PorterDuffXfermode(mode));
                }
                canvas.drawBitmap(topLayer, new Matrix(), paint);
                topLayer.recycle();
            }
        }
        System.gc();

        return baseLayer;
    }

    /**
     * @param b
     * @return
     */
    public static boolean bitmapExists(@Nullable Bitmap b) {
        return b != null && !b.isRecycled();
    }

    public static void destroyBitmap(@NonNull Bitmap b) {
        if (bitmapExists(b)) {
            b.recycle();
        }
    }

    @Nullable
    public static Bitmap getBitmapFromVector(
            @NonNull Context context, int drawableId, @NonNull BitmapFactory.Options options) {
        Drawable drawable = ContextCompat.getDrawable(context, drawableId);
        if(drawable == null) { return null; }

        int width = drawable.getIntrinsicWidth() == 0 ?
                options.outWidth : drawable.getIntrinsicWidth();
        width = Math.max(width, options.outWidth);

        int height = drawable.getIntrinsicHeight() == 0 ?
                options.outHeight : drawable.getIntrinsicHeight();
        height = Math.max(height, options.outHeight);

        Log.d("BitmapDims",
                width + " " + height + "->" +
                        options.outWidth + " " + options.outHeight + "->" +
                        drawable.getIntrinsicWidth() + " " + drawable.getIntrinsicHeight());

        Bitmap bitmap = Bitmap.createBitmap(
                width,
                height,
                Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0,
                width,    //use dimensions of Drawable
                height
        );

        drawable.draw(canvas);
        return bitmap;
    }

}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/utilities/BitmapUtils.kt
package com.TritiumGaming.phasmophobiaevidencepicker.data.utilities

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.util.Log
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat
import javax.microedition.khronos.egl.EGL10
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.egl.EGLContext
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.max

/**
 * BitmapUtils class
 *
 * @author TritiumGamingStudios
 */
class BitmapUtils {
    var resources: ArrayList<Int> = ArrayList()
    private var currentLayer = -1

    private var filters: ArrayList<PorterDuff.Mode?> = ArrayList()

    private var maxTextureSize = 0

    init {
        setMaxTextureSize()
    }

    fun setResource(@DrawableRes resource: Int): BitmapUtils {
        clearResources()
        addResource(resource)

        return this
    }

    private fun addResource(@DrawableRes resource: Int): BitmapUtils {
        resources.add(resource)
        filters.add(null)

        return this
    }

    fun addResource(@DrawableRes resource: Int, filterMode: PorterDuff.Mode?): BitmapUtils {
        resources.add(resource)
        filters.add(filterMode)

        return this
    }

    fun clearResources() {
        resources = ArrayList()
        filters = ArrayList()
        currentLayer = -1
    }

    private fun setMaxTextureSize() {
        // Safe minimum default size
        val IMAGE_MAX_BITMAP_DIMENSION = 2048

        // Get EGL Display
        val egl = EGLContext.getEGL() as EGL10
        val display = egl.eglGetDisplay(EGL10.EGL_DEFAULT_DISPLAY)

        // Initialise
        val version = IntArray(2)
        egl.eglInitialize(display, version)

        // Query total number of configurations
        val totalConfigurations = IntArray(1)
        egl.eglGetConfigs(display, null, 0, totalConfigurations)

        // Query actual list configurations
        val configurationsList = arrayOfNulls<EGLConfig>(
            totalConfigurations[0]
        )
        egl.eglGetConfigs(
            display,
            configurationsList,
            totalConfigurations[0],
            totalConfigurations
        )

        val textureSize = IntArray(1)
        var maximumTextureSize = 0

        // Iterate through all the configurations to located the maximum texture size
        for (i in 0 until totalConfigurations[0]) {
            // Only need to check for width since opengl textures are always squared
            egl.eglGetConfigAttrib(
                display,
                configurationsList[i],
                EGL10.EGL_MAX_PBUFFER_WIDTH,
                textureSize
            )

            // Keep track of the maximum texture size
            if (maximumTextureSize < textureSize[0]) {
                maximumTextureSize = textureSize[0]
            }
        }
        // Release
        egl.eglTerminate(display)

        maxTextureSize = max(maximumTextureSize.toDouble(), IMAGE_MAX_BITMAP_DIMENSION.toDouble())
            .toInt()
    }

    fun compileBitmaps(context: Context): Bitmap? {

        var bitmap: Bitmap? = null

        for (i in resources.indices) {
            val drawableRes = resources[i]
            val filter = filters[i]

            bitmap = createBitmap(context, bitmap, drawableRes, filter)
        }
        return bitmap
    }

    /**
     * @param previousBitmap Can be null. If bitmap is null, an empty layer will propagate
     * in its place. If not null, the bitmap will be the layer.
     */
    fun compileNextBitmap(context: Context, previousBitmap: Bitmap?): Bitmap? {
        currentLayer++
        return createBitmap(context, previousBitmap, resources[currentLayer], null)
    }

    fun hasNextBitmap(): Boolean {
        return currentLayer < resources.size - 1
    }

    private fun createBitmap(
        c: Context,
        baseLayer: Bitmap?,
        drawableResId: Int,
        mode: PorterDuff.Mode?
    ): Bitmap? {
        val options = BitmapFactory.Options()
        options.inSampleSize = 1
        options.inJustDecodeBounds = true

        // Raw height and width of image with respect to starting sampleSize
        val height = options.outHeight
        val width = options.outWidth
        //Get biggest image dimension between both width and height
        val highestDim = max(height.toDouble(), width.toDouble()).toInt()
        val dimScale = maxTextureSize.toDouble() / highestDim.toDouble() * options.inDensity
        if (dimScale < 1) {
            options.inSampleSize += ceil(abs(dimScale)).toInt()
        }
        options.inJustDecodeBounds = false

        var bitmapToAdd = BitmapFactory.decodeResource(c.resources, drawableResId, options)
        if (bitmapToAdd == null) {
            //Temp Options
            val optionsTemp = BitmapFactory.Options()
            optionsTemp.inSampleSize = 1
            optionsTemp.inJustDecodeBounds = true
            if (baseLayer != null) {
                optionsTemp.outWidth = baseLayer.width
                optionsTemp.outHeight = baseLayer.height
            }
            optionsTemp.inSampleSize = options.inSampleSize
            optionsTemp.inJustDecodeBounds = false

            bitmapToAdd = getBitmapFromVector(c, drawableResId, optionsTemp)
        }

        return addLayer(baseLayer, bitmapToAdd!!, mode)
    }

    @Throws(OutOfMemoryError::class)
    private fun addLayer(baseLayer: Bitmap?, topLayer: Bitmap, mode: PorterDuff.Mode?): Bitmap? {
        var baseLayer = baseLayer
        if (baseLayer == null && bitmapExists(topLayer)) {
            baseLayer = Bitmap.createBitmap(
                topLayer.width,
                topLayer.height,
                topLayer.config
            )
        }
        if (baseLayer != null && !baseLayer.isRecycled) {
            val canvas = Canvas(baseLayer)
            if (bitmapExists(topLayer)) {
                var paint: Paint? = null
                if (mode != null) {
                    paint = Paint()
                    paint.setXfermode(PorterDuffXfermode(mode))
                }
                canvas.drawBitmap(topLayer, Matrix(), paint)
                topLayer.recycle()
            }
        }
        System.gc()

        return baseLayer
    }

    companion object {
        @JvmStatic
        fun bitmapExists(b: Bitmap?): Boolean {
            return b != null && !(b.isRecycled)
        }

        fun destroyBitmap(b: Bitmap?) {
            if (bitmapExists(b)) {
                b?.recycle()
            }
        }

        fun getBitmapFromVector(
            context: Context, drawableId: Int, options: BitmapFactory.Options
        ): Bitmap? {
            val drawable = ContextCompat.getDrawable(context, drawableId) ?: return null

            var width =
                if (drawable.intrinsicWidth == 0) options.outWidth else drawable.intrinsicWidth
            width = max(width.toDouble(), options.outWidth.toDouble()).toInt()

            var height =
                if (drawable.intrinsicHeight == 0) options.outHeight else drawable.intrinsicHeight
            height = max(height.toDouble(), options.outHeight.toDouble()).toInt()

            Log.d(
                "BitmapDims",
                width.toString() + " " + height + "->" +
                        options.outWidth + " " + options.outHeight + "->" +
                        drawable.intrinsicWidth + " " + drawable.intrinsicHeight
            )

            val bitmap = Bitmap.createBitmap(
                width,
                height,
                Bitmap.Config.ARGB_8888
            )
            val canvas = Canvas(bitmap)
            drawable.setBounds(
                0, 0,
                width,  //use dimensions of Drawable
                height
            )

            drawable.draw(canvas)
            return bitmap
        }
    }
}
